zcm/julia/ZCM.jl

```diff
@@ -15,320 +35,253 @@ end
+        pointer = ccall(("zcm_create", "libzcm"), Ptr{Native.Zcm}, (Cstring,), provider);
+        instance = new(pointer, provider, Subscription[])
+        finalizer(instance, close)
+        return instance
     end
+end
 
-    function Zcm() return Zcm(""); end
+function close(zcm::Zcm)
+    if zcm.zcm != C_NULL
+        ccall(("zcm_destroy", "libzcm"), Void,
+              (Ptr{Native.Zcm},), zcm)
+        zcm.zcm = C_NULL
+    end
 end
-export Zcm;
 
-function encode end
-function decode end
+# Defines the conversion when we pass this to a C function expecting a pointer
+Base.unsafe_convert(::Type{Ptr{Native.Zcm}}, zcm::Zcm) = zcm.zcm
 
-function good(zcm::Zcm)
-    return (zcm.zcm != C_NULL) && (errno(zcm) == 0);
-end
-export good;
+good(zcm::Zcm) = (zcm.zcm != C_NULL) && (errno(zcm) == 0)
 
-function errno(zcm::Zcm)
-    return ccall(("zcm_errno", "libzcm"), Cint, (Ptr{Native.Zcm},), zcm.zcm);
-end
-export errno;
+errno(zcm::Zcm) = ccall(("zcm_errno", "libzcm"), Cint, (Ptr{Native.Zcm},), zcm)
 
 function strerror(zcm::Zcm)
-    val =  ccall(("zcm_strerror", "libzcm"), Cstring, (Ptr{Native.Zcm},), zcm.zcm);
+    val =  ccall(("zcm_strerror", "libzcm"), Cstring, (Ptr{Native.Zcm},), zcm)
     if (val == C_NULL)
-        return "unable to get strerror";
+        return "unable to get strerror"
     else
-        return unsafe_string(val);
+        return unsafe_string(val)
     end
 end
-export strerror;
 
-# TODO: get this into docs:
-# Note to self: handler could be either a function or a functor, so long as it has
-#               handler(rbuf::RecvBuf, channel::String, usr) defined
-function subscribe(zcm::Zcm, channel::AbstractString, handler, usr)
-
-    handler_jl = __zcm_handler(handler, usr);
-    handler_c  = cfunction(__zcm_handler_wrapper, Void,
-                           (Ptr{Native.RecvBuf}, Cstring, Ptr{Void},));
-    uv_wrapper = ccall(("uv_zcm_msg_handler_create", "libzcmjulia"), Ptr{Native.UvSub},
-                       (Ptr{Void}, Any,), handler_c, handler_jl);
-    uv_handler = cglobal(("uv_zcm_msg_handler_trigger", "libzcmjulia"));
-
-    return Sub(ccall(("zcm_subscribe", "libzcm"), Ptr{Native.Sub},
-                     (Ptr{Native.Zcm}, Cstring, Ptr{Void}, Ptr{Native.UvSub},),
-                     zcm.zcm, channel, uv_handler, uv_wrapper),
-               handler_jl, handler_c, uv_handler, uv_wrapper);
+function onresponse(rbuf::Native.RecvBuf, channelbytes::Cstring, opts::SubscriptionOptions)
+    channel = unsafe_string(channelbytes)
+    msgdata = unsafe_wrap(Vector{UInt8}, rbuf.data, rbuf.data_size)
+    msg = decode(opts.msgtype, msgdata)
+    opts.handler(channel, msg)
+    return nothing
+end
+
+function subscribe(zcm::Zcm, channel::AbstractString, options::T) where T <: SubscriptionOptions
+    # The C function is called with the receive buffer, channel name, and the 
+    # SubscriptionOptions struct
+    c_handler = cfunction(onresponse, Void, Tuple{Ref{Native.RecvBuf}, Cstring, Ref{T}})
```

Comment:
Why a `Tuple` here? This seems inconsistent with a couple lines below where we use the `()` syntax

Reply:
either appears to work, but I think you're right that `()` is what's in the manual. Resolved in 9fcd25f